### ISSUE TITLE
[network] Propagate ProtocolId to application layer

### ISSUE THREAD

**Issue by zekun000** [libra/libra#2486]
Today we dispatch network messages based on protocol id statically, while there's need to dispatch the messages based on the on-chain state.

We want to avoid exposing the on-chain state to network, we need to push the protocol id higher to application layer and delegate the dispatch to application.
```
match notif {
  NetworkNotification::RecvMessage(peer_id, msg) => {
    if msg.protocol != onChainConfig.consensus_protocol() {
      epoch_mgr.process_different_protocol(msg.protocol);
    } else {
      match msg.protocol {
        HotStuff => {
          let msg = HotStuffMsg::from(msg.mdata)?;
          hotstuff.process_msg(msg);
        },
        PBFT => {
          let msg = PBFTMsg::from(msg.mdata)?;
          pbft.process_msg(msg);
        },
      }
    }
  }
}
```

**Comment by davidiw** [comment]
I wonder if this is something that network should support or if it should e part of the interfaces choice to return relevant metadata. In the current networking model, network owns this area where the components themselves only get component relevant information. More specifically, network has protocol ids, but a component might prefer to have its own notation. This is similar to how the new consensus LCS interface works. We could imagine the channel / context being provided beyond the networking scope.

**Comment by zekun000** [comment]
@davidiw I agree components can have its own metadata on top of the network protocol ids. Although I'd prefer to have `one` unified way to carry metadata instead of two, if we let each components carry the metadata, we don't really need protocol id in networking layer other than indicating which component it is?

**Comment by phlip9** [comment]
We have talked about a few things that might be helpful:

1) [`NewPeer`](https://github.com/libra/libra/blob/master/network/src/validator_network/mod.rs#L70) events should include the full (?) set of protocols.

Today, `NewPeer` events only contain the `PeerId` of the connecting peer. Instead, it could contain a `Vec<ProtocolId>` from the identity exchange.

2) After the QoS diff from @bothra90, `NetworkEvents` looks like: 

```
     Stream<PeerManagerNotification> --.
                                        > -- NetworkEvents<TMessage>: Stream<Event<TMessage>
Stream<ConnectionStatusNotification> --`
```

which automatically deserializes the raw bytes in `PeerManagerNotification` to the generic protobuf type `TMessage` and joins in the `NewPeer`/`LostPeer` events from `ConnectionStatusNotification`.

Instead, maybe `NetworkEvents` should not deserialize yet, since the deserialization would depend on the protocol version, or we could just return the raw component streams to consensus specifically?

Alternatively, we keep `NetworkEvents` the same and consensus registers separate handles for each version it supports?

3) Likewise, `NetworkSender` wraps a `PeerManagerRequestSender` under the hood. Maybe for consensus we just return a `PeerManagerRequestSender` for consensus' [`add_to_network`](https://github.com/libra/libra/blob/master/network/src/validator_network/consensus.rs#L50)?

For context, `PeerManagerRequestSender` thinks only in terms of bytes, i.e., `PeerManagerRequestSender::send_to` takes a `(PeerId, ProtocolId, Bytes)`, whereas `NetworkSender` takes a `(PeerId, ProtocolId, TMessage)` and serializes the message before passing it down.

Ideally, we come up with a sustainable API from consensus to network. Maybe consensus is a special enough case that we just give the "raw" `PeerManagerRequestSender` and `Receiver<PeerManagerNotification>` channels. I'm interested in what you think would be most helpful.

**Comment by phlip9** [comment]
> I wonder if this is something that network should support or if it should e part of the interfaces choice to return relevant metadata. In the current networking model, network owns this area where the components themselves only get component relevant information. More specifically, network has protocol ids, but a component might prefer to have its own notation. This is similar to how the new consensus LCS interface works. We could imagine the channel / context being provided beyond the networking scope.

Personally, I would rather us have a more unified system than a bunch of ad-hoc versioning schemes.

**Comment by davidiw** [comment]
I'm curious why a single network endpoint would route to potentially multiple different endpoints both outgoing or incoming. In your example, @zekun000, it would seem more prudent to have a HotstuffConsensusNetwork and a PbftConsensusNetwork and not even bother with muxing. If we were to support upgradability, I think networking and other systems would be easiest configured by allowing the system to dynamically change the receiver or handler of events.

I think it is much safer to use code / static routing / compiler rather than receive-time conditional / interpreted. Maybe we should chat more about this. At the same time, it is absolutely trivial to pass back the protocol.

@phlip9, per the conversation in the most recent set of PRs. I agree and think it is valuable to push the validator_network code closer to where the types are since that area should inherently depend on networking and if not, we can always put networking under a feature flag. So you could imagine having 3 files:
* NetworkHandler, specialized to NetworkTypes
* NetworkTypes
* Binding between network and subsystem

Alternatively we could just merge NetworkTypes and NetworkHandler into the same file.

Thoughts?
